x/shield/keeper/pool.go

```diff
@@ -29,134 +111,67 @@ func (k Keeper) GetPool(ctx sdk.Context, id uint64) (types.Pool, bool) {
+	// Update foreign service fees in the global pool.
+	totalServiceFees := k.GetServiceFees(ctx)
+	totalServiceFees = totalServiceFees.Add(types.MixedDecCoins{Foreign: sdk.NewDecCoinsFromCoins(serviceFees.Foreign...)})
+	k.SetServiceFees(ctx, totalServiceFees)
 
 	return pool, nil
 }
 
-// UpdatePool updates pool info and sponsor's shield.
-func (k Keeper) UpdatePool(ctx sdk.Context, updater sdk.AccAddress, shield sdk.Coins, deposit types.MixedCoins, id uint64, addTime time.Duration, description string) (types.Pool, error) {
+// UpdatePool updates pool info and shield for B.
+func (k Keeper) UpdatePool(ctx sdk.Context, poolID uint64, description string, updater sdk.AccAddress, shield sdk.Coins, serviceFees types.MixedCoins) (types.Pool, error) {
 	admin := k.GetAdmin(ctx)
 	if !updater.Equals(admin) {
 		return types.Pool{}, types.ErrNotShieldAdmin
 	}
 
-	// Check if shield is backed by admin's delegations.
-	provider, found := k.GetProvider(ctx, admin)
-	if !found {
-		return types.Pool{}, types.ErrNoDelegationAmount
-	}
-	shieldAmt := shield.AmountOf(k.sk.BondDenom(ctx))
-	provider.Collateral = provider.Collateral.Add(shieldAmt)
-	if shieldAmt.GT(provider.Available) {
-		return types.Pool{}, sdkerrors.Wrapf(types.ErrInsufficientStaking,
-			"available %s, shield %s", provider.Available, shield)
-	}
-	provider.Available = provider.Available.Sub(shieldAmt)
-
-	pool, found := k.GetPool(ctx, id)
+	// Update pool info.
+	pool, found := k.GetPool(ctx, poolID)
 	if !found {
 		return types.Pool{}, types.ErrNoPoolFound
 	}
-	pool.EndTime = pool.EndTime.Add(addTime)
-	pool.TotalCollateral = pool.TotalCollateral.Add(shieldAmt)
-	poolCertiKCollateral, found := k.GetPoolCertiKCollateral(ctx, pool)
-	if !found {
-		poolCertiKCollateral = types.NewCollateral(pool, admin, sdk.ZeroInt())
-	}
-	poolCertiKCollateral.Amount = poolCertiKCollateral.Amount.Add(shieldAmt)
-	pool.Shield = pool.Shield.Add(shield...)
-	pool.Premium = pool.Premium.Add(types.MixedDecCoinsFromMixedCoins(deposit))
 	if description != "" {
 		pool.Description = description
 	}
+	k.SetPool(ctx, pool)
 
-	// Transfer deposit and store.
-	if err := k.DepositNativePremium(ctx, deposit.Native, admin); err != nil {
-		return types.Pool{}, err
-	}
-
-	// Update sponsor purchase.
-	sponsorPurchase, found := k.GetPurchaseList(ctx, id, pool.SponsorAddr)
-
-	purchaseEndTime := pool.EndTime.Add(-k.gk.GetVotingParams(ctx).VotingPeriod * 2)
-	if purchaseEndTime.Before(ctx.BlockTime()) {
-		return types.Pool{}, types.ErrPoolLifeTooShort
+	// Update purchase and shield.
+	totalServiceFees := k.GetServiceFees(ctx)
+	if !shield.IsZero() {
+		if _, err := k.PurchaseShield(ctx, poolID, shield, "shield for sponsor", updater, serviceFees.Native); err != nil {
+			return pool, err
+		}
+	} else if !serviceFees.Native.IsZero() {
+		// Allow adding service fees without purchasing more shield.
+		totalServiceFees = totalServiceFees.Add(types.MixedDecCoins{Foreign: sdk.NewDecCoinsFromCoins(serviceFees.Foreign...)})
```

Comment:
Did you mean adding native DecCoins? Also, I think it should be added to `totalServiceFees.Native`?

Reply:
Yeah should be native.